### PR TITLE
Implement Cache Committee for RPC Server

### PR DESCRIPTION
**Pull request by terenc3t**
**Problem:**
Currently for every RPC assignment call beacon node recalculates crosslink committee list of that given slot. A huge performance upgrade is to cache these crosslink committees list. Which means we only need one calculation per slot, save it to the crosslink committee cache for RPC server to respond.

**Cache structure:"**
In short, the committees cache is defined as below, **key** is the slot, and **value** is the committee list. Using the following structure because it's the most-used fields and it'll give us the highest performance upgrade.
**slot**: `1` -> **committees**: `[[shard:1, committees[1,2,3]], [shard:2, committees[4,5,6]]]`
**slot**: `2` -> **committees**: `[[shard:3, committees[7,8,9]], [shard:4, committees[10,11]]]`
**slot**: `3` -> **committees**: `[[shard:5, committees[12,13]], [shard:6, committees[14,15,16]]]`

**Todos:**
- [x] implement crosslink committee cache
- [x] use cache in RPC validator server
- [x] fix existing regression tests
- [x] implement new crosslink committee cache tests
- [x] implement tests in validator server to use this caches
- [x] implement feature flag
- [x] test run time works with 1 node and 8 validators
- [x] test run time works with 3 nodes and 3, 3 and 2 validators
- [x] look at cache hit/miss ratio
- [x] deploy to cluster

### CODE REVIEW

**Comment by prestonvanloon** [comment]
Request: please add a feature flag to enable this. `--enable-committee-cache` or similar

File: beacon-chain/rpc/validator_server.go [hunk 1]
```diff
@@ -150,26 +162,66 @@ func (vs *ValidatorServer) CommitteeAssignment(
...
+	for slot := req.EpochStart; slot < req.EpochStart+params.BeaconConfig().SlotsPerEpoch; slot++ {
+		if featureconfig.FeatureConfig().EnableCommitteesCache {
+			if exists, committees, err := vs.committeesCache.CommitteesInfoBySlot(int(slot)); exists || err != nil {
+				if err != nil {
+					return nil, err
+				}
+				span.AddAttributes(trace.BoolAttribute("committeeCacheHit", true))
+				committee, shard, slot, isProposer, err :=
+					helpers.ValidatorAssignment(index, slot, committees.committees)
+				if err == nil {
```
Comment (nisdas) [inline beacon-chain/rpc/validator_server.go]: what of err != nil, the error isnt returned here

**Comment by terenc3t** [comment]
@nisdas reorg generally doesn't hurt assignments, reorg only hurts head block and head state. the assignment only matters from current epoch and next epoch.

For example:
Current slot is 80, for some reason we had a reorg and head slot becomes 78. The assignment should still fetch slot 80 - 88, the assignment won't change because of head went back two slots.

File: beacon-chain/rpc/validator_server.go (hunk 1 above)
Comment (terenc3t) [reply beacon-chain/rpc/validator_server.go]: the only error is `fmt.Errorf("unable to find assignment for validator %d at slot %d", vIndex, slot)` that just means cache miss and we'll add it to the cache below

File: beacon-chain/powchain/service.go [hunk 2]
```diff
@@ -247,7 +247,7 @@ func (w *Web3Service) processSubscribedHeaders(header *gethTypes.Header) {
 	blockNumberGauge.Set(float64(header.Number.Int64()))
 	w.blockHeight = header.Number
 	w.blockHash = header.Hash()
-	w.blockTime = time.Unix(int64(header.Time), 0)
+	//w.blockTime = time.Unix(int64(header.Time), 0)
```
Comment (prestonvanloon) [inline beacon-chain/powchain/service.go]: revert this please

File: beacon-chain/rpc/validator_server.go [hunk 3]
```diff
@@ -4,6 +4,7 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"github.com/prysmaticlabs/prysm/shared/featureconfig"
```
Comment (prestonvanloon) [inline beacon-chain/rpc/validator_server.go]: goimports?

File: beacon-chain/rpc/validator_server.go [hunk 4]
```diff
@@ -140,6 +144,9 @@ func (vs *ValidatorServer) assignment(
 	epochStart uint64,
 ) (*pb.CommitteeAssignmentResponse_CommitteeAssignment, error) {
 
+	_, span := trace.StartSpan(ctx, "beacon-chain.rpcservice.CommitteeAssignment.assignment")
```
Comment (prestonvanloon) [inline beacon-chain/rpc/validator_server.go]: why not replace ctx? without this, the span isn't passed to children.

```suggestion
	ctx, span := trace.StartSpan(ctx, "beacon-chain.rpcservice.CommitteeAssignment.assignment")
```

**Comment by terenc3t** [comment]
Don't merge this in please, I found some regression while running 3 nodes with 3,3,2 validator set up. This is more likely due to bad merge. I'll work on resolving this now.

Regression error:
`[2019-04-06 07:03:42] ERROR validator: Failed to update assignments error=rpc error: code = Unknown desc = could not find validator 7 assignment starting epoch 17`

**Pull request merged**